src/virtio-net.c

```diff
@@ -0,0 +1,292 @@
+        return;
+    }
+    vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+    return;
+}
+
+void virtio_net_complete_request_tx(struct virtq *vq) {
+    struct virtio_net_dev *dev = (struct virtio_net_dev *)vq->dev;
+    vm_t *v = container_of(dev, vm_t, virtio_net_dev);
+    struct vring_packed_desc *desc;
+    while((desc = virtq_get_avail(vq)) != NULL){
+        uint8_t *data = vm_guest_to_host(v, desc->addr);
+        size_t virtio_header_len = sizeof(struct virtio_net_hdr_v1);
+
+        if (desc->len < virtio_header_len) {
+            vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+            return;
+        }
+        
+        uint8_t *actual_data = data + virtio_header_len;
+        size_t actual_data_len = desc->len - virtio_header_len;
+
+        struct iovec iov[1];
+        iov[0].iov_base = actual_data;
+        iov[0].iov_len = actual_data_len;
+
+        ssize_t write_bytes = writev(dev->tapfd, iov, 1);
+        if (write_bytes < 0) {
+            vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+            return;
+        }
+        desc->flags ^= (1ULL << VRING_PACKED_DESC_F_USED);
+        dev->virtio_pci_dev.config.isr_cap.isr_status |= VIRTIO_PCI_ISR_QUEUE;
+        return;
+    }
+    vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+    return;
+}
+
+static struct virtq_ops virtio_net_rx_ops = {
+    .enable_vq = virtio_net_enable_vq_rx,
+    .complete_request = virtio_net_complete_request_rx, 
+    .notify_used = virtio_net_notify_used_rx,
+};
+
+static struct virtq_ops virtio_net_tx_ops = {
+    .enable_vq = virtio_net_enable_vq_tx,
+    .complete_request = virtio_net_complete_request_tx, 
+    .notify_used = virtio_net_notify_used_tx,
+};
+
+bool virtio_net_init(struct virtio_net_dev *virtio_net_dev){
+    memset(virtio_net_dev, 0x00, sizeof(struct virtio_net_dev));
+
+    virtio_net_dev->tapfd = open("/dev/net/tun", O_RDWR);
+    if(virtio_net_dev->tapfd < 0){
+        return false;
+    }
+    struct ifreq ifreq = {.ifr_flags = IFF_TAP | IFF_NO_PI};
+    strncpy(ifreq.ifr_name, TAP_INTERFACE, sizeof(ifreq.ifr_name));
+    if (ioctl(virtio_net_dev->tapfd, TUNSETIFF, &ifreq) < 0) {
+        fprintf(stderr, "failed to allocate TAP device: %s\n", strerror(errno));
+        return false;
+    }
+    assert(fcntl(virtio_net_dev->tapfd, F_SETFL,
+                fcntl(virtio_net_dev->tapfd, F_GETFL, 0) | O_NONBLOCK) >= 0);
+    return true;
+}
+static void virtio_net_setup(struct virtio_net_dev *dev)
+{
+    vm_t *v = container_of(dev, vm_t, virtio_net_dev);
+
+    dev->enable = true;
+    dev->irq_num = VIRTIO_NET_IRQ;
+    dev->rx_ioeventfd = eventfd(0, EFD_CLOEXEC);
+    dev->tx_ioeventfd = eventfd(0, EFD_CLOEXEC);
+    dev->irqfd = eventfd(0, EFD_CLOEXEC);
+    vm_irqfd_register(v, dev->irqfd, dev->irq_num, 0);
+    for (int i = 0; i < VIRTIO_NET_VIRTQ_NUM; i++){
+        struct virtq_ops *ops = (i == 0) ? &virtio_net_rx_ops : &virtio_net_tx_ops;
```

Comment:
I'll suggest to declare an array for `rx_ops` and `tx_ops` and iterate here.

```cpp
#define VIRTQ_RX 0
#define VIRQ_TX 1

static struct virtq_ops virtio_net_ops[VIRTIO_NET_VIRTQ_NUM] = {
    [VIRTQ_RX] = {
        .enable_vq = virtio_net_enable_vq_rx,
       .complete_request = virtio_net_complete_request_rx,·
       .notify_used = virtio_net_notify_used_rx,
    },
    [VIRQ_TX] = {
        .enable_vq = virtio_net_enable_vq_tx,
       .complete_request = virtio_net_complete_request_tx,·
       .notify_used = virtio_net_notify_used_tx,
    },
}
```

Reply:
Yes, I have already made the changes as you suggested.